Idris won't find libgmp when linking compiled binaries

Let's try to compile something simple as in example —
https://eb.host.cs.st-andrews.ac.uk/Idris/first.html
(containing main : IO (); main = putStrLn "Hello world";)

\# idris hello.idr -o hello
gcc: rawSystem: runInteractiveProcess: exec: does not exist (No such file or directory)

Okay, let's install gcc —
\# idris hello.idr -o hello
/nix/store/kxf1p7l7lgm6j5mjzkiwcwzc98s9f1az-binutils-2.23.1/bin/ld: cannot find -lgmp
collect2: error: ld returned 1 exit status
FAILURE: "gcc" ["-O2","-fwrapv","-fno-strict-overflow","-DHAS_PTHREAD","-DIDRIS_ENABLE_STATS","-I.","-x","c","/tmp/idris7198853861649760492","-L/nix/store/3x07l4v8wv8msri8r7w5lhb8hg4vw734-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/rts","-lidris_rts","-lgmp","-lpthread","-I/nix/store/3x07l4v8wv8msri8r7w5lhb8hg4vw734-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/rts","-lm","-I.","-I.","-I/nix/store/3x07l4v8wv8msri8r7w5lhb8hg4vw734-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/prelude","-I/nix/store/3x07l4v8wv8msri8r7w5lhb8hg4vw734-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/base","-o","hello"]

Now installed gmp, but it doesn't matter —
\# idris hello.idr -o hello
/nix/store/kxf1p7l7lgm6j5mjzkiwcwzc98s9f1az-binutils-2.23.1/bin/ld: cannot find -lgmp
[...the same as in previous command...]

What can i do to compile (at least) hello.idr?

Idris REPL is almost ok except some things related to the "outer world", such as putStrLn.
emacs interface is ok.

I tried to export all paths that i found in your link...
The same thing ;-(
% idris ./hello.idr -o hello                                                      
/nix/store/v7h3j43vx0dz5ahhkxg5z50by2iqc6k1-binutils-2.23.1/bin/ld: cannot find -lgmp
collect2: error: ld returned 1 exit status
FAILURE: "gcc" ["-O2","-fwrapv","-fno-strict-overflow","-DHAS_PTHREAD","-DIDRIS_ENABLE_STATS","-I.","-x","c","/tmp/idris20448977631967513926","-L/nix/store/j9kvp0rbg4hs4xjilg2ppl613pmx69q3-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/rts","-lidris_rts","-lgmp","-lpthread","-I/nix/store/j9kvp0rbg4hs4xjilg2ppl613pmx69q3-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/rts","-lm","-I.","-I.","-I/nix/store/j9kvp0rbg4hs4xjilg2ppl613pmx69q3-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/prelude","-I/nix/store/j9kvp0rbg4hs4xjilg2ppl613pmx69q3-idris-0.9.19.1/share/x86_64-linux-ghc-7.10.2/idris-0.9.19.1/base","-o","hello"]

% idris -V
[...]Version 0.9.20[...]

Problem remains the same (gmp installed) —
% idris hello.idr -o hello
/nix/store/v7h3j43vx0dz5ahhkxg5z50by2iqc6k1-binutils-2.23.1/bin/ld: cannot find -lgmp
[...]

--- a/pkgs/development/idris-modules/default.nix
+++ b/pkgs/development/idris-modules/default.nix
@@ -1,4 +1,4 @@
-{ pkgs, idris, overrides ? (self: super: {}) }: let
+{ pkgs, idris-no-deps, overrides ? (self: super: {}) }: let
   inherit (pkgs.lib) callPackageWith fix' extends;
 
   /* Taken from haskell-modules/default.nix, should probably abstract this away */
@@ -33,7 +33,12 @@
 
     value = callPackage (./. + "/${name}.nix") {};
   }) files)) // {
-    inherit idris callPackage;
+    inherit idris-no-deps callPackage;
+    # See #10450 about why we have to wrap the executable
+    idris =
+        (pkgs.callPackage ./idris-wrapper.nix {})
+          idris-no-deps
+          { path = [ pkgs.gcc ]; lib = [pkgs.gmp]; };
 
     # A list of all of the libraries that come with idris
     builtins = pkgs.lib.mapAttrsToList (name: value: value) builtins_;

--- a/pkgs/development/idris-modules/idris-wrapper.nix
+++ b/pkgs/development/idris-modules/idris-wrapper.nix
@@ -1,0 +1,14 @@
+{ symlinkJoin, makeWrapper, stdenv }: idris: { path, lib }:
+
+symlinkJoin {
+  name = idris.name;
+  src = idris.src;
+  paths = [ idris ];
+  buildInputs = [ makeWrapper ];
+  postBuild = ''
+    wrapProgram $out/bin/idris \
+      --suffix PATH : ${ stdenv.lib.makeBinPath path } \
+      --suffix LIBRARY_PATH : ${stdenv.lib.makeLibraryPath lib}
+      '';
+  }
+

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -5876,7 +5876,8 @@ with pkgs;
   icedtea_web = icedtea8_web;
 
   idrisPackages = callPackage ../development/idris-modules {
-    idris =
+
+    idris-no-deps =
       let
         inherit (self.haskell) lib;
         haskellPackages = self.haskellPackages.override {
@@ -5892,6 +5893,8 @@ with pkgs;
       in
         haskellPackages.idris;
   };
+
+  idris = idrisPackages.with-packages [ idrisPackages.base ] ;
 
   intercal = callPackage ../development/compilers/intercal { };
 

